email-service/cmd/main.go

```diff
@@ -0,0 +1,238 @@
+	"github.com/GenesisEducationKyiv/software-engineering-school-4-0-Hukyl/pkg/settings"
+)
+
+const defaultCronSpec = "0 0 12 * * *"
+
+const emailTimeout = 5 * time.Second
+
+var appConfig appCfg.Config
+
+func InitDatabase() (*database.DB, error) {
+	db, err := database.New(dbCfg.NewFromEnv())
+	if err != nil {
+		return nil, err
+	}
+	if err := db.Migrate(&models.Rate{}, &models.Subscriber{}); err != nil {
+		return nil, err
+	}
+	return db, nil
+}
+
+func InitNotificationsCron(db *database.DB, mailer notifications.EmailClient) *cron.Manager {
+	// Start cron job for notifications
+	spec := appConfig.NotificationCropSpec
+	if spec == "" {
+		spec = defaultCronSpec
+	}
+
+	notifier := notifications.NewMailNotifier(
+		mailer,
+		models.NewRateRepository(db),
+		models.NewSubscriberRepository(db),
+		&message.PlainRate{},
+	)
+
+	notifyF := func() error {
+		ctx, cancel := context.WithTimeout(context.Background(), emailTimeout)
+		defer cancel()
+		return notifier.Notify(ctx)
+	}
+
+	cronManager := cron.NewManager()
+	err := cronManager.AddJob(spec, notifyF)
+	if err != nil {
+		slog.Error("adding job", slog.Any("error", err))
+	}
+	return cronManager
+}
+
+func doWithContext(ctx context.Context, f func() error) error {
+	done := make(chan struct{})
+	go func() {
+		defer close(done)
+		if err := f(); err != nil {
+			slog.Error("error", slog.Any("error", err))
+		}
+	}()
+
+	for {
+		select {
+		case <-ctx.Done():
+			return ctx.Err()
+		case <-done:
+			return nil
+		}
+	}
+}
+
+func InitRateConsumer(config transportCfg.Config, rateRepo *models.RateRepository) *rate.Client {
+	rateConsumer, err := rate.NewClient(transportCfg.Config{
+		QueueName: appConfig.RateQueueName,
+		BrokerURI: config.BrokerURI,
+	})
+	if err != nil {
+		slog.Error("creating rate client", slog.Any("error", err))
+		return nil
+	}
+	rateConsumer.Subscribe(func( // nolint: errcheck
+		ctx context.Context,
+		from, to string, rate float32, time time.Time,
+	) error {
+		slog.Info(
+			"rate fetched",
+			slog.Any("from", from),
+			slog.Any("to", to),
+			slog.Any("rate", rate),
+			slog.Any("time", time),
+		)
+		rateModel := &models.Rate{
+			CurrencyFrom: from,
+			CurrencyTo:   to,
+			Rate:         rate,
+			Created:      time.Unix(),
+		}
+		err := doWithContext(ctx, func() error {
+			return rateRepo.Create(rateModel)
+		})
+		if err != nil {
+			slog.Error("saving rate", slog.Any("error", err))
+			return err
+		}
+		return nil
+	})
+	return rateConsumer
+}
+
+func InitSubscriberConsumer(
+	config transportCfg.Config, subRepo *models.SubscriberRepository,
+) *subscriber.Client {
+	subConsumer, err := subscriber.NewClient(transportCfg.Config{
+		QueueName: appConfig.UserQueueName,
+		BrokerURI: config.BrokerURI,
+	})
+	if err != nil {
+		slog.Error("creating subscriber client", slog.Any("error", err))
+		return nil
+	}
+	err = subConsumer.SubscribeCreate(func(ctx context.Context, email string) error {
+		slog.Info("new subscriber", slog.Any("email", email))
```

Comment:
I don't think this is a good place to write business logic. main.go should be as clean/small as possible

see https://github.com/GenesisEducationKyiv/software-engineering-school-4-0-Hukyl/pull/9/files#r1669747729